### PR TITLE
Allow module level skip

### PR DESCRIPTION
**Pull request by georgeyk**
Work-in-progress fix for https://github.com/pytest-dev/pytest/issues/2805
I need to update the PR with all contributing guideline recommendations, but a initial feedback would be good.

--//--
Thanks for submitting a PR, your contribution is really appreciated!

Here's a quick checklist that should be present in PRs:

- [x] Add a new news fragment into the changelog folder
  * name it `$issue_id.$type` for example (588.bug)
  * if you don't have an issue_id change it to the pr id after creating the pr
  * ensure type is one of `removal`, `feature`, `bugfix`, `vendor`, `doc` or `trivial`
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."
- [x] Target: for `bugfix`, `vendor`, `doc` or `trivial` fixes, target `master`; for removals or features target `features`;
- [x] Make sure to include reasonable tests for your change if necessary

Unless your change is a trivial or a documentation fix (e.g.,  a typo or reword of a small section) please:

- [x] Add yourself to `AUTHORS`;

### CODE REVIEW

File: _pytest/outcomes.py [hunk 1]
```diff
@@ -62,14 +62,21 @@ def exit(msg):
...
 
-def skip(msg=""):
+def skip(msg="", **kwargs):
     """ skip an executing test with the given message.  Note: it's usually
     better to use the pytest.mark.skipif marker to declare a test to be
     skipped under certain conditions like mismatching platforms or
     dependencies.  See the pytest_skipping plugin for details.
+
+    :kwarg bool allow_module_level: allows this function to be called at
+    module level, skipping the rest of the module. Default to False.
```
Comment (nicoddemus) [inline _pytest/outcomes.py]: I think this needs to be indented otherwise `tox -e docs` fails

File: _pytest/outcomes.py [hunk 2]
```diff
@@ -62,14 +62,21 @@ def exit(msg):
...
     """
     __tracebackhide__ = True
-    raise Skipped(msg=msg)
+    allow_module_level = kwargs.pop('allow_module_level', False)
+    if kwargs:
+        keys = [k for k in kwargs.keys()]
+        raise TypeError('unexpected keyworkd arguments: {}'.format(keys))
```
Comment (nicoddemus) [inline _pytest/outcomes.py]: Needs to use `{0}` to work on Python 2.6:

```python
raise TypeError('unexpected keyworkd arguments: {0}'.format(keys))
```

**Comment by The-Compiler** [comment]
This probably should go to the `features` branch.

File: _pytest/outcomes.py (hunk 2 above)
Comment (The-Compiler) [reply _pytest/outcomes.py]: keyworkd -> keyword

File: doc/en/skipping.rst [hunk 3]
```diff
@@ -54,6 +54,15 @@ by calling the ``pytest.skip(reason)`` function:
         if not valid_config():
             pytest.skip("unsupported configuration")
 
+It is also possible to skip the whole module using
+``pytest.skip(reason, allow_module_level=True)`` at the module level:
+
+
+.. code-block:: python
+
```
Comment (RonnyPfannschmidt) [inline doc/en/skipping.rst]: i do wonder if there is a better example for this (thats not looking like a skipif marker or a importorskip would already solve

Comment (georgeyk) [reply doc/en/skipping.rst]: Maybe using the use case from previous discussion ? Something like:
```
if not pytest.config.getoption('--custom-flag'):
    pytest.skip('--custom-flag is missing, skipping tests', allow_module_level=True)
```

**Comment by nicoddemus** [comment]
You can rebase on the `features` branch, push force to this branch and then edit the target branch of the PR here on GH to `features`.

**Comment by georgeyk** [comment]
Is this `event.when` supposed to exist in CollectReport instance ?  https://github.com/pytest-dev/pytest/blob/features/_pytest/skipping.py#L352

**Comment by nicoddemus** [comment]
Hmm `folded_skips` is relatively new I think. @RonnyPfannschmidt do you have a suggestion on how to proceed here?

**Comment by RonnyPfannschmidt** [comment]
nope

**Comment by RonnyPfannschmidt** [comment]
actually i do, whole module skips should be ignored, collectrports have no when

so at collect time we should never fold skips

**Comment by nicoddemus** [comment]
> so at collect time we should never fold skips

I see, thanks. Could you tell @georgeyk what should be changed then? I'm not sure what should be done with collectreports.

And the error is happening at the end of the session, when we are about to show the terminal summary, not during collection:

```
Traceback (most recent call last):
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/_pytest/pytester.py", line 744, in runpytest_inprocess
    reprec = self.inline_run(*args, **kwargs)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/_pytest/pytester.py", line 717, in inline_run
    ret = pytest.main(list(args), plugins=plugins)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/_pytest/config.py", line 59, in main
    return config.hook.pytest_cmdline_main(config=config)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 745, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 339, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 334, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 614, in execute
    res = hook_impl.function(*args)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/_pytest/main.py", line 131, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/_pytest/main.py", line 125, in wrap_session
    exitstatus=session.exitstatus)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 745, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 339, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 302, in __call__
    return outcome.get_result()
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 279, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 265, in __init__
    self.result = func()
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 300, in <lambda>
    outcome = _CallOutcome(lambda: self.oldcall(hook, hook_impls, kwargs))
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 334, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 613, in execute
    return _wrapped_call(hook_impl.function(*args), self.execute)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 250, in _wrapped_call
    wrap_controller.send(call_outcome)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/_pytest/terminal.py", line 427, in pytest_sessionfinish
    exitstatus=exitstatus)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 745, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 339, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 302, in __call__
    return outcome.get_result()
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 279, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 265, in __init__
    self.result = func()
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 300, in <lambda>
    outcome = _CallOutcome(lambda: self.oldcall(hook, hook_impls, kwargs))
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 334, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/pluggy.py", line 614, in execute
    res = hook_impl.function(*args)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/_pytest/skipping.py", line 291, in pytest_terminal_summary
    show_skipped(terminalreporter, lines)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/_pytest/skipping.py", line 370, in show_skipped
    fskips = folded_skips(skipped)
  File "/home/travis/build/pytest-dev/pytest/.tox/py36/lib/python3.6/site-packages/_pytest/skipping.py", line 352, in folded_skips
    if event.when == 'setup' and 'skip' in keywords and 'pytestmark' not in keywords:
AttributeError: 'CollectReport' object has no attribute 'when'
```

**Comment by RonnyPfannschmidt** [comment]
correct - a skip in a collect-report is added to the skip stats, and such a stat shouldn't be folded in the display

**Comment by RonnyPfannschmidt** [comment]
one solution could be to add a when property to collect reports

another solution could be to just ignore them

its not clear what has the best effect

**Comment by nicoddemus** [comment]
> correct - a skip in a collect-report is added to the skip stats, and such a stat shouldn't be folded in the display

That was my first thought too, but after thinking some more I think it might make sense in some cases. For example, if (say) 3 whole modules are skipped for the same reason, it makes sense to fold it.

In light of that it might make sense to just make a special case for it in `folded_skips`.

**Comment by nicoddemus** [comment]
Something like (untested):

```python
when = getattr(event, 'when', None)  # might be a collection report
if when == 'setup' and 'skip' in keywords and 'pytestmark' not in keywords:
```

**Comment by georgeyk** [comment]
@nicoddemus Thank you all to support this great project. I hope to do more in the future!

**Comment by nicoddemus** [comment]
> I hope to do more in the future!

We really appreciate it, thanks!

**Comment by nicoddemus** [comment]
@RonnyPfannschmidt you think we can merge this?

**Pull request merged**